src/app/components/login/login.component.html

```diff
@@ -2,27 +2,74 @@
   <div class="row justify-content-center mt-5">
     <h2 class="text-center mb-4">Bienvenido:</h2>
     <form>
-      <div class="form-group">
-        <label>Usuario:</label>
+      <mat-form-field class="example-full-width">
+        <mat-label>Usuario</mat-label>
         <input
           type="text"
-          class="form-control"
-          id="usuario"
-          placeholder="Ingresa tu usuario"
+          matInput
+          [formControl]="userFormControl"
+          [errorStateMatcher]="matcher"
+          placeholder="Ex. JoseManuel23"
         />
-      </div>
-      <div class="form-group mt-3">
-        <label>Contraseña:</label>
+        <mat-hint>Usuario valido</mat-hint>
+        <mat-error
+          *ngIf="
+            userFormControl.hasError('minlength') &&
+            userFormControl.hasError('pattern') &&
+            !userFormControl.hasError('required')
+          "
+        >
+          Por favor inserta un usuario válido.
+        </mat-error>
+        <mat-error *ngIf="userFormControl.hasError('required')">
+          Usuario <strong> requerido</strong>
+        </mat-error>
+        <mat-error *ngIf="userFormControl.hasError('minlength')">
+          Usuario demasido <strong>corto</strong>
+        </mat-error>
+        <mat-error *ngIf="userFormControl.hasError('pattern')">
+          El usuario no permite <strong>caracteres especiales</strong>
+        </mat-error>
```

Comment:
This is the same code, repeated 3 times, refactor into one `<mat-error>` with dinamic text